src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionChangeService.cs

```diff
@@ -0,0 +1,428 @@
+                return new LocationOptionMetaChange
+                {
+                    DataSetVersionId = nextVersionId,
+                    PreviousState = null,
+                    CurrentState = LocationOptionMetaChange.State.Create(target)
+                };
+            });
+
+        publicDataDbContext.LocationMetaChanges.AddRange(locationMetaChanges);
+        publicDataDbContext.LocationOptionMetaChanges.AddRange([
+            .. locationOptionMetaDeletionsAndChanges, .. locationOptionMetaAdditions
+        ]);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private async Task CreateGeographicLevelChange(
+        Guid previousVersionId,
+        Guid nextVersionId,
+        CancellationToken cancellationToken)
+    {
+        var oldGeographicLevelMeta = await GetGeographicLevelMeta(previousVersionId, cancellationToken);
+        var newGeographicLevelMeta = await GetGeographicLevelMeta(nextVersionId, cancellationToken);
+
+        var levelsAreEqual =
+            newGeographicLevelMeta.Levels.Order().SequenceEqual(oldGeographicLevelMeta.Levels.Order());
+
+        if (!levelsAreEqual)
+        {
+            publicDataDbContext.GeographicLevelMetaChanges.Add(new GeographicLevelMetaChange
+            {
+                DataSetVersionId = nextVersionId,
+                PreviousStateId = oldGeographicLevelMeta.Id,
+                CurrentStateId = newGeographicLevelMeta.Id
+            });
+            await publicDataDbContext.SaveChangesAsync(cancellationToken);
+        }
+    }
+
+    private async Task CreateIndicatorChanges(
+        Guid previousVersionId,
+        Guid nextVersionId,
+        CancellationToken cancellationToken)
+    {
+        var oldIndicatorMetas = await GetIndicatorMetas(previousVersionId, cancellationToken);
+        var newIndicatorMetas = await GetIndicatorMetas(nextVersionId, cancellationToken);
+
+        var indicatorMetaDeletions = oldIndicatorMetas
+            .Except(newIndicatorMetas, IndicatorMeta.PublicIdComparer)
+            .OrderBy(indicatorMeta => indicatorMeta.Label)
+            .Select(indicatorMeta => new IndicatorMetaChange
+            {
+                DataSetVersionId = nextVersionId,
+                PreviousStateId = indicatorMeta.Id,
+                CurrentStateId = null
+            })
+            .ToList();
+
+        var indicatorMetaAdditions = newIndicatorMetas
+            .Except(oldIndicatorMetas, IndicatorMeta.PublicIdComparer)
+            .OrderBy(indicatorMeta => indicatorMeta.Label)
+            .Select(indicatorMeta => new IndicatorMetaChange
+            {
+                DataSetVersionId = nextVersionId,
+                PreviousStateId = null,
+                CurrentStateId = indicatorMeta.Id
+            })
+            .ToList();
+
+        publicDataDbContext.IndicatorMetaChanges.AddRange([.. indicatorMetaDeletions, .. indicatorMetaAdditions]);
```

Comment:
As discussed in meeting - we should handle changes to indicators which are automatic as well i.e. updates to labels, units, decimal places.

Reply:
done